### PR TITLE
[ticket/6466] Use jQuery and custom dropdown for permission roles tooltips

### PR DESCRIPTION
**Pull request by marc1706**
Ticket: https://tracker.phpbb.com/browse/PHPBB3-6466

PHPBB3-6466

### CODE REVIEW

**Comment by marc1706** [comment]
Including video: https://www.m-a-styles.de/download/tooltips.ogv

File: phpBB/adm/style/tooltip.js [hunk 1]
```diff
@@ -10,7 +10,12 @@ phpBB Development Team:
 	- further adjustements
 */
 
-var head_text, tooltip_mode;
+(function($) { // Avoid conflicts with other libraries
+
+"use strict";
```
Comment (callumacrae) [inline phpBB/adm/style/tooltip.js]: single quotes

File: phpBB/adm/style/tooltip.js [hunk 2]
```diff
@@ -66,24 +71,25 @@ function enable_tooltips_select(id, headline, sub_id) {
...
-
-	document.getElementsByTagName('body')[0].appendChild(hold);
+	$('body').append(hold);
 
 	if (id === null) {
-		links = document.getElementsByTagName('option');
+		$links = $('.roles-options li');
 	} else {
-		links = document.getElementById(id).getElementsByTagName('option');
+		$links = $('#' + id + ' .roles-options li');
```
Comment (callumacrae) [inline phpBB/adm/style/tooltip.js]: This would be more efficient as `$('.roles-options li', '#' + id)`

File: phpBB/adm/style/tooltip.js [hunk 3]
```diff
@@ -164,52 +170,49 @@ function create_element(tag, c) {
 * Correct positioning of tooltip container
 */
 function locate(e) {
```
Comment (callumacrae) [inline phpBB/adm/style/tooltip.js]: `e` generally means "error", could you change this to `$el` at the least?

Comment (marc1706) [inline phpBB/adm/style/tooltip.js]: Sure, just didn't modify it for now. Still slightly working on the overall behavior.

File: phpBB/adm/style/tooltip.js [hunk 4]
```diff
@@ -10,206 +10,173 @@ phpBB Development Team:
...
-	tooltip_mode = 'link';
-}
+var tooltips = [];
 
 /**
-* Enable tooltip replacements for selects
+ * Enable tooltip replacements for selects
+ * @param {string} id ID tag of select
+ * @param {string} headline Text that should appear on top of tooltip
+ * @param {string} subId Sub ID that should only be using tooltips (optional)
```
Comment (callumacrae) [inline phpBB/adm/style/tooltip.js]: For optional params, wrap the name in square brackets:

`@param {string} [subId] Sub ID that should only be using tooltips`

http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values

File: phpBB/adm/style/tooltip.js [hunk 5]
```diff
@@ -10,206 +10,173 @@ phpBB Development Team:
...
-* Create new element
+ * Create new element
+ *
+ * @param {string} tag HTML tag
+ * @param {string} c Element's class
+ *
+ * @return {object} Created element
 */
-function create_element(tag, c) {
+phpbb.createElement = function (tag, c) {
```
Comment (callumacrae) [inline phpBB/adm/style/tooltip.js]: This already exists in jQuery!

```js
$('<div />', { 'class': 'top' });
```

Slightly more verbose, but also readable. I had no idea what the create_element function did

**Comment by callumacrae** [comment]
I don't really care that much, but there's some pretty long lines which could easily be made shorter and so more readable :smile:

File: phpBB/adm/style/tooltip.js (hunk 5 above)
Comment (marc1706) [inline phpBB/adm/style/tooltip.js]: But probably also slower :)

Comment (callumacrae) [inline phpBB/adm/style/tooltip.js]: By 2.608 nanoseconds! http://jsperf.com/jquery-create-vs-native-element

That's not a difference worth caring about :P

http://blog.methvin.com/2013/12/please-stop-jsperfcom-abuse.html

**Comment by marc1706** [comment]
@Nicofuma I'd say this is good to go now

**Comment by callumacrae** [comment]
js lgtm

**Pull request merged**